src/overmap.cpp

```diff
@@ -3406,133 +3417,44 @@ void overmap::place_rifts(int const z)
     }
 }
 
-struct node
-{
-    int x;
-    int y;
-    int d;
-    int p;
-
-    node( int xp, int yp, int dir, int pri ) {
-        x = xp; y = yp; d = dir; p = pri;
-    }
-    // Operator overload required by priority queue interface.
-    bool operator< ( const node &n ) const {
-        return this->p > n.p;
-    }
-};
-
-void overmap::make_hiway(int x1, int y1, int x2, int y2, int z, const std::string &base)
+void overmap::make_hiway( int x1, int y1, int x2, int y2, int z, const std::string &base )
 {
-    if (x1 == x2 && y1 == y2) {
-        return;
-    }
-
-    std::priority_queue<node, std::deque<node> > nodes[2];
-    bool closed[OMAPX][OMAPY] = {{false}};
-    int open[OMAPX][OMAPY] = {{0}};
-    int dirs[OMAPX][OMAPY] = {{0}};
-    int dx[4] = {1, 0, -1, 0};
-    int dy[4] = {0, 1, 0, -1};
-    int i = 0;
-    int disp = (base == "road") ? 5 : 2;
-
-    nodes[i].push(node(x1, y1, 5, 1000));
-    open[x1][y1] = 1000;
+    const tripoint source( x1, y1, z );
+    const tripoint dest( x2, y2, z );
+    const int disp = base == "road" ? 5 : 2;
+
+    const auto estimate = [ this, disp, &base, &dest ]( const pf::node &prev, const pf::node &cur ) {
+        // Reject nodes that don't allow roads to cross them (e.g. buildings)
+        if( !road_allowed( ter( cur.x, cur.y, dest.z ) ) ) {
+            return -1;
+        }
+        // Reject nodes that make corners on the river
+        if( prev.dir != cur.dir && ( is_river( ter( prev.x, prev.y, dest.z ) ) ||
+                                     is_river( ter( cur.x, cur.y, dest.z ) ) ) ) {
+            return -1;
+        }
+
+        int res = ( std::abs( dest.x - cur.x ) + std::abs( dest.y - cur.y ) ) / disp;
+        // Prefer existing roads.
+        res += check_ot_type( base, cur.x, cur.y, dest.z ) ? 0 : 3;
+        // Prefer flat land over bridges
+        res += !is_river( ter( cur.x, cur.y, dest.z ) ) ? 0 : 2;
+        // Try not to turn too much
+        //res += (mn.d == d) ? 0 : 1;
```

Comment:
For some reason this line was commented out in `overmap::make_hiway()` and I kept it intact.
It's supposed to slightly penalize direction changes to make roads less curvy. What shall we do?